### DISCUSSION TITLE
Batching load problem

### DISCUSSION THREAD

**Discussion by wolfjiang** [graphql-java/graphql-java#3212] [General]
**Describe the bug**
I have query like, queryStr:
```
{
     Customer{
         id
         name
        primaryContacts: contacts("isPrimary='t'){id, name}
        otherContacts: contacts("isPrimary='f'){id, code, name}
    }
}
```

I will update my graphql server from graphql-java V7.0 to graphql-java V22
The old  contacts data fetcher using @batched annotation and can get sources from environment fine, and primayrContacts and otherContacts query as two batch.

when using V22, I refacter code as:

```
BatchLoaderWithContext batchLoader = BatchLoaderWithContext<Map<String,Object>, Object> {
  @Override
  public CompletionStage<List<Object>> load(List<Map<String,Object>> sources, BatchLoaderEnvironment environment) {
    return CompletableFuture.supplyAsync(() -> {
      DataFetchingEnvironment env = (DataFetchingEnvironment)environment.getKeyContextsList().get(0);  //---TODO1
      return fetchFromDBWithEnv(sources, env);
    });
  }
}

GraphQL graphql = GraphQL.newGraphQL(schema)
        .queryExecutionStrategy(new AsyncExecutionStrategy()).build();

DataLoaderRegistry registry = new DataLoaderRegistry();
registry.register("batchLoader", DataLoaderFactory.newDataLoader(batchLoader));

ExecutionInput.Builder builder = ExecutionInput.newExecutionInput()
        .dataLoaderRegistry(registry)
        .query(queryStr);
```
and the contacts data fetcher :
```
contactsDataFetcher = new  DataFetcher<Object> {
  public Object get(DataFetchingEnvironment environment) {
    DataLoader<Object, Object> dataLoader = environment.getDataLoader("batchLoader");
    return dataLoader.load(environment.getSource(), environment)); 
 } 
}
```
There are some question at --TODO1 above:
1) primaryContacts and otherContacts query same field will merge togother or not ?
1) debug set breakpoint at TODO1,  found the code enter TODO1 only once, and only the first context get. as the result  primaryContacts and primaryContacts has same data
2) at TODO1, each source has a DataFetchingEnvironment, if Customer has 3 records, then sources size is 3, and KeyContexts from BatchLoaderEnvironment has 3 too, and also 3 DataFetchingEnvironment is corresponds to 3 souces. but i only need the field argumens and the select fields. how to merge DataFetchingEnvironment if primaryContacts and otherContacts are merged in one dataloader?  Or can i using first DataFetchingEnvironment when primaryContacts and otherContacts are sperate batch?

Hope to receive a response from the expert， Thanks!

**Comment by bbakerman** [comment]
I am moving this to discussion as its really a question not a bug

**Comment by bbakerman** [comment]
> primaryContacts and otherContacts query same field will merge togother or not ?

No - these have alises on them - they are different result fields

> debug set breakpoint at TODO1, found the code enter TODO1 only once, and only the first context get. as the result primaryContacts and primaryContacts has same data

Right so is a batch call - so you should have 1 call with 2 keys in it.

I am not sure why you do this

```
environment.getKeyContextsList().get(0)
```

There will be two different context items here - one for each dataloader call


> at TODO1, each source has a DataFetchingEnvironment, if Customer has 3 records, then sources size is 3, and KeyContexts from BatchLoaderEnvironment has 3 too, and also 3 DataFetchingEnvironment is corresponds to 3 souces. but i only need the field argumens and the select fields. how to merge DataFetchingEnvironment if primaryContacts and otherContacts are merged in one dataloader? Or can i using first DataFetchingEnvironment when primaryContacts and otherContacts are sperate batch?

You cant merge them - they are different fields - the data loader works differently to the old `@Batched` which did kinda merge together fields executions.
